File: random.cpp
#include "random.h"
#include "strlib.h"

#define BASE 10
#define LENGTH 8
#define MAX_STATE 100000

char toUpper(char input){
  if(input>95)return input-32;
  else return input;
}

Random::Random(std::string input){
  state=0;
  hash=sha256(input);
}
int Random::generate(){
  int ret = 0;
  int arr[LENGTH];
  for (int i=0;i<LENGTH;i++){
    arr[i]=0;
  }
  for (int i=0;i<64;i++){
    arr[i%LENGTH]+=hash[i]%BASE;
  }
  for (int i=0;i<LENGTH;i++){
    ret=ret*BASE+arr[i];
  }
  hash=sha256(hash+intToString(getState())+intToString(ret));

  return ret;
}
int Random::getState(){
  state=(state+1)%MAX_STATE;
  return state;
}

File: JS/random.js
const BASE = 10;
const LENGTH = 8;
const MAX_STATE = 100000;

function sha256(pwd){
  var crypto = require('crypto');
  var hash = crypto.createHash('sha256').update(pwd).digest('hex');
  return hash;
}
class Random{
  constructor(input) {
    this.state = 0;
    this.hash = sha256(input);
  }
  getState(){
    this.state = (this.state + 1) % MAX_STATE;
    return this.state;
  }
  generate(){
    var ret = 0;
    var arr= new Array(LENGTH).fill(0);
    for (var i = 0; i < 64; i++){
        arr[i % LENGTH] += this.hash.charCodeAt(i) % BASE;
    }
    for (var i = 0; i < LENGTH; i++){
        ret = ret * BASE + arr[i];
    }
    this.hash = sha256(this.hash + this.getState() + ret);
    return ret;
  }
}
module.exports = {
  Random : Random
}
